File: app/src/main/java/gex/com/masinqojava/SongAdapter.java
package gex.com.masinqojava;

import android.content.Context;
import android.media.MediaMetadataRetriever;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import java.io.IOException;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import com.bumptech.glide.Glide;
import java.util.ArrayList;

public class SongAdapter extends RecyclerView.Adapter {


    private final Context context;
    private final ArrayList<SongTemplate> songs;


    SongAdapter(ArrayList<SongTemplate> songs, Context context) {
        this.songs = songs;
        this.context = context;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.audio_items_template, parent, false);
        return new SongViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {

        SongViewHolder songViewHolder = (SongViewHolder) holder;
        songViewHolder.songTitle.setText(songs.get(position).getTitle());
        songViewHolder.songArtist.setText(songs.get(position).getArtist());
        songViewHolder.songDuration.setText(songs.get(position).getDuration());
        byte[] image = getAlbumArt(songs.get(position).getPath());
        if (image != null){
            Glide.with(context).asBitmap()
                    .load(image)
                    .into(songViewHolder.album_art);
        }
        else {
            Glide.with(context).asBitmap()
                    .load(R.drawable.ic_launcher_background)
                    .into(songViewHolder.album_art);
        }
    }

    public void clear(){
        songs.clear();
        notifyDataSetChanged();
    }
    @Override
    public int getItemCount() {
        return songs.size();
    }

    public static class SongViewHolder extends RecyclerView.ViewHolder {

        TextView songTitle;
        TextView songArtist;
        TextView songDuration;
        ImageView album_art;
        public SongViewHolder(@NonNull View itemView) {
            super(itemView);
            songTitle = itemView.findViewById(R.id.audio_title);
            songArtist = itemView.findViewById(R.id.audio_artist);
            album_art = itemView.findViewById(R.id.audio_icon);
            songDuration = itemView.findViewById(R.id.audio_duration);
        }
    }
    private byte[] getAlbumArt(String uri){

        MediaMetadataRetriever metadataRetriever = new MediaMetadataRetriever();
            try {
            metadataRetriever.setDataSource(uri);
            byte[] art = metadataRetriever.getEmbeddedPicture();
            metadataRetriever.release();
            return art;
        }catch (IOException e){
            e.printStackTrace();
            return null;
        }
    }
}

File: app/src/main/java/gex/com/masinqojava/SongAdapter.kt
package gex.com.masinqojava

import android.content.ContentUris
import android.content.Context
import android.media.MediaMetadataRetriever
import android.net.Uri
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import java.io.IOException
import java.util.Locale
import androidx.core.net.toUri
import androidx.media3.common.MediaItem

class SongAdapter internal constructor(
    private val songs: ArrayList<MediaItem?>,
    private val context: Context,
    private val onItemClick: (position: Int) -> Unit
) : RecyclerView.Adapter<SongAdapter.SongViewHolder>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SongViewHolder {
        val view =
            LayoutInflater.from(context).inflate(R.layout.audio_items_template, parent, false)
        return SongViewHolder(view)
    }

    override fun onBindViewHolder(holder: SongViewHolder, position: Int) {
        val song = songs[position]?: return
        val metadata = song.mediaMetadata
        holder.songTitle.text = metadata.title
        holder.songArtist.text = metadata.artist
        holder.songDuration.text = formatDuration(metadata.durationMs?.toInt())

        Glide.with(context)
            .asBitmap()
            .load(metadata.artworkUri)
            .thumbnail(
                Glide.with(context)
                    .asBitmap()
                    .load(R.drawable.ic_launcher_background)
                    .centerCrop()
            )
            .placeholder(R.drawable.ic_launcher_background)
            .error(R.drawable.ic_launcher_foreground)
            .centerCrop()
            .into(holder.albumArt)
        holder.itemView.setOnClickListener {
            onItemClick(position)
        }
    }

    fun clear() {
        songs.clear()
        notifyDataSetChanged()
    }

    override fun getItemCount(): Int {
        return songs.size
    }

    class SongViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var songTitle: TextView = itemView.findViewById<TextView>(R.id.audio_title)
        var songArtist: TextView = itemView.findViewById<TextView>(R.id.audio_artist)
        var albumArt: ImageView = itemView.findViewById<ImageView>(R.id.audio_icon)
        var songDuration: TextView = itemView.findViewById<TextView>(R.id.audio_duration)
    }


    private fun formatDuration(rawDuration: Int?): String {
        val seconds = (rawDuration?.div(1000))?.rem(60)
        val minutes = (rawDuration?.div((1000 * 60)))?.rem(60)
        return String.format(Locale.ROOT, "%d:%02d", minutes, seconds)
    }
}
